File: src/validations/palette.ts
import { z } from "zod"

export const hexColorSchema = z
  .string()
  .trim()
  .regex(/^#?[0-9A-Fa-f]{6}$/, "Invalid hex color")
  .transform((value) => {
    const normalized = value.toUpperCase()
    return normalized.startsWith("#") ? normalized : `#${normalized}`
  })

export const generatePaletteSchema = z.object({
  prompt: z.string().trim().min(3).max(500),
  mood: z
    .enum(["warm", "cool", "vibrant", "muted", "neutral", "dark", "light"])
    .optional(),
  style: z
    .enum([
      "minimal",
      "retro",
      "nature",
      "tech",
      "luxury",
      "playful",
      "corporate",
      "artistic",
    ])
    .optional(),
  color_count: z.number().int().min(3).max(10).optional(),
  base_colors: z.array(hexColorSchema).max(10).optional(),
})

const importedCommentSchema = z.object({
  text: z.string().trim().min(1).max(2000),
  id: z.string().min(1).optional(),
  imageId: z.string().optional(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
})

export const importedPaletteSchema = z.object({
  name: z.string().trim().min(1).max(120).optional(),
  colors: z.array(hexColorSchema).min(1).max(20),
  tagIds: z.array(z.string().min(1)).max(50).optional(),
  groupIds: z.array(z.string().min(1)).max(50).optional(),
  comments: z.array(importedCommentSchema).max(50).optional(),
})

export function parseImportedPalette(input: unknown) {
  return importedPaletteSchema.safeParse(input)
}

File: backend/app/schemas/palette.py
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


class PaletteMood(str, Enum):
    WARM = "warm"
    COOL = "cool"
    VIBRANT = "vibrant"
    MUTED = "muted"
    NEUTRAL = "neutral"
    DARK = "dark"
    LIGHT = "light"


class PaletteStyle(str, Enum):
    MINIMAL = "minimal"
    RETRO = "retro"
    NATURE = "nature"
    TECH = "tech"
    LUXURY = "luxury"
    PLAYFUL = "playful"
    CORPORATE = "corporate"
    ARTISTIC = "artistic"


class GeneratePaletteRequest(BaseModel):
    prompt: str = Field(
        min_length=3,
        max_length=500,
        description="Describe the palette you want (mood, brand, scene, etc.)",
    )
    mood: Optional[PaletteMood] = None
    style: Optional[PaletteStyle] = None
    color_count: int = Field(default=5, ge=3, le=10)
    base_colors: List[str] = Field(default_factory=list, max_length=5)

    @field_validator("base_colors")
    @classmethod
    def validate_hex_colors(cls, colors: List[str]) -> List[str]:
        validated: List[str] = []
        for color in colors:
            normalized = color.strip().upper()
            if not normalized.startswith("#"):
                normalized = f"#{normalized}"
            if len(normalized) != 7:
                raise ValueError(f"Invalid hex color: {color}")
            validated.append(normalized)
        return validated


class GeneratedPalette(BaseModel):
    name: str
    colors: List[str]
    description: str
    mood: Optional[str] = None
    style: Optional[str] = None
    source: str = Field(description="ai or fallback")


class GeneratePaletteResponse(BaseModel):
    palette: GeneratedPalette


class HealthResponse(BaseModel):
    status: str
    ai_enabled: bool
